axelor-sale/src/main/java/com/axelor/apps/sale/service/SaleOrderGeneratorServiceImpl.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Axelor Business Solutions
+ *
+ * Copyright (C) 2005-2024 Axelor (<http://axelor.com>).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.axelor.apps.sale.service;
+
+import com.axelor.apps.base.AxelorException;
+import com.axelor.apps.base.db.Partner;
+import com.axelor.apps.base.service.CompanyService;
+import com.axelor.apps.sale.db.SaleOrder;
+import com.axelor.apps.sale.db.repo.SaleOrderRepository;
+import com.axelor.apps.sale.service.app.AppSaleService;
+import com.axelor.apps.sale.service.saleorder.SaleOrderInitValueService;
+import com.axelor.apps.sale.service.saleorder.SaleOrderOnChangeService;
+import com.axelor.inject.Beans;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+
+public class SaleOrderGeneratorServiceImpl implements SaleOrderGeneratorService {
+  SaleOrderRepository saleOrderRepository;
+  protected AppSaleService appSaleService;
+  protected CompanyService companyService;
+
+  @Inject
+  public SaleOrderGeneratorServiceImpl(
+      SaleOrderRepository saleOrderRepository,
+      AppSaleService appSaleService,
+      CompanyService companyService) {
+    this.saleOrderRepository = saleOrderRepository;
+    this.appSaleService = appSaleService;
+    this.companyService = companyService;
+  }
+
+  @Transactional(rollbackOn = {Exception.class})
+  @Override
+  public SaleOrder createSaleOrder(Partner clientPartner)
+      throws AxelorException, JsonProcessingException {
+    SaleOrder saleOrder = new SaleOrder();
+    boolean isTemplate = false;
+    SaleOrderInitValueService saleOrderInitValueService =
+        Beans.get(SaleOrderInitValueService.class);
+    saleOrderInitValueService.setIsTemplate(saleOrder, isTemplate);
+    saleOrderInitValueService.getOnNewInitValues(saleOrder);
+    saleOrder.setClientPartner(clientPartner);
```

Comment:
Before setting the clientPartner, please check if the partner respect the condition of the domain, you can look at SaleOrderDomainService#getPartnerBaseDomain, if not, throw an exception.